### PR TITLE
Bulk inserts

### PR DESCRIPTION
**Pull request by edefimov**
This is an updated version of PR #682 solving issue #1392  refactored for php 7.1 and with small improvements like bulk insert execution even when the number of rows exceeds database limits.

### CODE REVIEW

**Comment by Ocramius** [comment]
@deeky666 this patch introduces some fixups on your existing bulk-insert patch.

Specifically, it will force transactions when an `insert` operation is acting on a range of values bigger than what the database supports.

We should consider further scenarios too:

 * when a query string is bigger than what is endorsed by the DB engine
 * when the number of bound parameters is bigger than what is allowed by the DB engine (for example, that's 1000 on OracleDB)

File: lib/Doctrine/DBAL/Query/BulkInsertQuery.php [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+        if ($insertMaxRows > 0 && $totalValues > $insertMaxRows && !$this->connection->isTransactionActive()) {
+            throw QueryException::transactionRequired();
+        }
+
+        $offset = 0;
+        $result = 0;
+        while ($offset < $totalValues) {
+            $values = array_slice($this->values, $offset, $insertMaxRows);
+            $parameters = array_slice($this->parameters, $offset, $insertMaxRows);
```
Comment (LisTo4ek) [inline lib/Doctrine/DBAL/Query/BulkInsertQuery.php]: Here should be something like:
$parameters = array_slice($this->parameters, $offset * count($this->values), $insertMaxRows * count($this->values));

**Review by morozov — CHANGES_REQUESTED** [review]
@edefimov thank you for the patch. Please rebase it and make sure the most generic cases are tested in a functional way.

**Comment by karousn** [comment]
any news about this ? available in next major version(3) ?

**Comment by claudiobizzotto** [comment]
Until this gets incorporated into the next major release, here's a temporary option: https://github.com/pharako/mysql-dbal

**Comment by takahashiyuya** [comment]
i hope can use it.

**Comment by morozov** [comment]
Closing as stale.

**Pull request closed without merging**